code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/process/RegionMonitor.java

```diff
@@ -0,0 +1,85 @@
+import io.cattle.platform.object.process.ObjectProcessManager;
+import io.cattle.platform.object.util.DataAccessor;
+import io.cattle.platform.servicediscovery.service.RegionService;
+import io.cattle.platform.servicediscovery.service.impl.RegionUtil;
+import io.cattle.platform.servicediscovery.service.impl.RegionUtil.ExternalProject;
+import io.cattle.platform.servicediscovery.service.impl.RegionUtil.ExternalRegion;
+import io.cattle.platform.task.Task;
+import io.github.ibuildthecloud.gdapi.condition.Condition;
+import io.github.ibuildthecloud.gdapi.condition.ConditionType;
+
+import java.util.Arrays;
+import java.util.List;
+
+import javax.inject.Inject;
+
+public class RegionMonitor extends AbstractJooqDao implements Task {
+    @Inject
+    ObjectProcessManager objectProcessManager;
+    @Inject
+    ObjectManager objectManager;
+    @Inject
+    JsonMapper jsonMapper;
+    @Inject
+    RegionService regionService;
+
+    @Override
+    public void run() {
+        List<Region> regions = objectManager.find(Region.class, REGION.REMOVED, new Condition(ConditionType.NULL));
+        if (regions.size() == 0) {
+            return;
+        }
+        List<AccountLink> accountLinks = objectManager.find(AccountLink.class, ACCOUNT_LINK.REMOVED, null, ACCOUNT_LINK.LINKED_REGION_ID, 
+            new Condition(ConditionType.NOTNULL));
+        List<String> invalidStates = Arrays.asList(CommonStatesConstants.REGISTERING, CommonStatesConstants.REMOVING, CommonStatesConstants.REMOVED);
+        for(AccountLink link : accountLinks) {
+            if (invalidStates.contains(link.getState())) {
+                continue;
+            }
+            // targetRegion not present
+            Region targetRegion = objectManager.loadResource(Region.class, link.getLinkedRegionId());
+            if(targetRegion == null) {
+                delete(link);
+                continue;
+            }
+            try {
+                // localRegion in targetRegion not present 
+                Region localRegion = objectManager.findAny(Region.class, REGION.LOCAL, true, REGION.REMOVED, null);
+                ExternalRegion externalRegion = RegionUtil.getExternalRegion(targetRegion, localRegion.getName(), jsonMapper);
+                if(externalRegion == null) {
+                    delete(link);
+                    continue;
+                }
+                // environment not present or its uuid changed 
+                ExternalProject externalProject = RegionUtil.getTargetProjectByName(targetRegion, link.getLinkedAccount(), jsonMapper);
+                String storedUUID = DataAccessor.fieldString(link, "uuid");
+                if(externalProject == null || (storedUUID!=null && !externalProject.getUuid().equals(storedUUID))) {
+                    delete(link);
+                }
+            } catch (Exception ex) {
+                throw new RuntimeException(String.format("Failed to monitor account link for [%d]", link.getId()), ex);
```

Comment:
we should never throw a runtime in a periodic task. Just log it, proceed with the loop. The next loop will retry